src/espIdf/unitTest/adapter.ts

```diff
@@ -0,0 +1,274 @@
+              unitTestAppUri,
+              idfTestitem.testName,
+              cancelToken
+            );
+            result
+              ? testRun.passed(test, Date.now() - startTime)
+              : testRun.failed(
+                  test,
+                  new TestMessage("Error in test"),
+                  Date.now() - startTime
+                );
+          } catch (error) {
+            testRun.failed(
+              test,
+              new TestMessage(error.message),
+              Date.now() - startTime
+            );
+          }
+          testRun.appendOutput(`Completed ${test.id}\r\n`);
+        }
+        test.children.forEach((t) => queue.push(t));
+      }
+      testRun.end();
+    };
+
+    this.unitTestController.createRunProfile(
+      "Run Tests",
+      TestRunProfileKind.Run,
+      runHandler,
+      true,
+      undefined,
+      false
+    );
+
+    this.unitTestController.resolveHandler = async (item: TestItem) => {
+      if (!item) {
+        const fileList = await this.getFileList();
+        this.components = await this.getTestComponents(fileList);
+        await this.loadTests(fileList);
+        return;
+      }
+      const espIdfTestItem = await this.getTestsForFile(item.uri);
+      for (const child of espIdfTestItem.children) {
+        const childItem = this.unitTestController.createTestItem(
+          child.id,
+          child.label,
+          child.uri
+        );
+        item.children.add(childItem);
+        idfTestData.set(childItem, child);
+      }
+      idfTestData.set(item, espIdfTestItem);
+      this.unitTestController.items.add(item);
+    };
+    context.subscriptions.push(this.unitTestController);
+  }
+
+  async getOrCreateFile(file: Uri) {
+    const existing = this.unitTestController.items.get(file.toString());
+    if (existing) {
+      return {
+        testItem: existing,
+        espIdfTestItem: idfTestData.get(existing) as EspIdfTestItem,
+      };
+    }
+    const testItem = this.unitTestController.createTestItem(
+      file.toString(),
+      file.fsPath.split("/").pop(),
+      file
+    );
+    testItem.canResolveChildren = true;
+    this.unitTestController.items.add(testItem);
+    const espIdfTestItem: EspIdfTestItem = {
+      type: "suite",
+      id: file.fsPath,
+      label: file.fsPath.split("/").pop(),
+      filePath: file.fsPath,
+      children: [],
+      uri: file,
+      testName: "TEST_ALL",
+    };
+    idfTestData.set(testItem, espIdfTestItem);
+    return { testItem, espIdfTestItem };
+  }
+
+  private async getFileList(): Promise<Uri[]> {
+    let files: Uri[] = [];
+
+    try {
+      const fileRegexStr = readParameter("idf.testFileRegex") as string;
+      files = await workspace.findFiles("**/test/test_*.c");
+    } catch (err) {
+      window.showErrorMessage("Cannot find test result path!", err);
+      return [];
+    }
+    return files;
+  }
+
+  private async getTestComponents(files: Uri[]): Promise<string[]> {
```

Comment:
There are a few modifications I would consider for better performance and readability, not sure if really necessary but:

1) We could use a Set instead of an Array for componentsList (we also rename this variable to componentsSet) and then convert if to an array when we return it with `Array.from(componentsSet)` so that we won't have any duplicated components and we can get rid of extra iteration over componentsList with `.includes()`.
2) Since we only iterate over the files, I think it would be better for readability to use `.forEach()` instead of `.map() `

```javascript
private async getTestComponents(files: Uri[]): Promise<string[]> {
  const componentsSet = new Set<string>();
  files.forEach((match) => {
    const componentName = basename(match.fsPath.split("/test/test_")[0]);
    componentsSet.add(componentName);
  });
  return Array.from(componentsSet);
}
```